### PR TITLE
add ssl support to webserver

### PR DESCRIPTION
**Pull request by b2un0**
### Motivation for changes:

- remote access to flexget

### Detailed changes:

- add ssl support to webserver.py (cherrypy)

### Config usage if relevant (new plugin or updated schema):
```
web_server:
  bind: 0.0.0.0
  port: 3539
  ssl_certificate: '/etc/ssl/private/myCert.pem'
  ssl_private_key: '/etc/ssl/private/myKey.key'
api: yes
webui: yes
```

### CODE REVIEW

**Comment by b2un0** [comment]
image

File: flexget/webserver.py [hunk 1]
```diff
@@ -36,6 +36,8 @@
             'properties': {
                 'bind': {'type': 'string', 'format': 'ipv4', 'default': '0.0.0.0'},
                 'port': {'type': 'integer', 'default': 3539},
+                'ssl_certificate': {'type': 'string', 'format': 'string', 'default': ''},
+                'ssl_private_key': {'type': 'string', 'format': 'string', 'default': ''},
```
Comment (paranoidi) [inline flexget/webserver.py]: I believe in schema you can tell that both of these are required or neither.

**Comment by stevezau** [comment]
Ah! this has been on the todo list for a while.. 

I think we should support the generation of a self signed certificate?

File: flexget/webserver.py (hunk 1 above)
Comment (b2un0) [inline flexget/webserver.py]: they are not required, if you don't set them, the webserver starts without ssl.

**Comment by b2un0** [comment]
`I think we should support the generation of a self signed certificate?`
self signed certificate even works

**Comment by Tommatheussen** [comment]
What I believe @stevezau meant is provide a way to the user that allows Flexget to create self-signed certificates

**Comment by liiight** [comment]
@stevezau if that is what you meant, allow flexget to create the ssl, I don't think we need that. 

I believe @paranoidi meant json schema dependencies: https://spacetelescope.github.io/understanding-json-schema/reference/object.html#dependencies

Specifically bidirectional dependencies.

File: flexget/webserver.py [hunk 2]
```diff
@@ -213,12 +219,22 @@ def _start_server(self):
...
+            })
+
         try:
             host = self.bind if self.bind != "0.0.0.0" else socket.gethostbyname(socket.gethostname())
         except socket.gaierror:
             host = '127.0.0.1'
 
-        log.info('Web interface available at http://%s:%s' % (host, self.port))
+        if self.ssl_certificate and self.ssl_private_key:
+            log.info('Web interface available at https://%s:%s' % (host, self.port))
```
Comment (cvium) [inline flexget/webserver.py]: Let the logger handle string formatting ie. `log.info('Web interface available at https://%s:%s' , host, self.port)`

**Comment by liiight** [comment]
Please set the schema to reflect a bidirectional dependency:
```
'depenedencies': {
  'ssl_certificate': ['ssl_private_key'],
  'ssl_private_key': ['ssl_certificate']
}
```

**Comment by b2un0** [comment]
dependencies was added

File: flexget/webserver.py (hunk 2 above)
Comment (b2un0) [inline flexget/webserver.py]: done

File: flexget/webserver.py [hunk 3]
```diff
@@ -232,9 +235,11 @@ def _start_server(self):
             host = '127.0.0.1'
 
         if self.ssl_certificate and self.ssl_private_key:
-            log.info('Web interface available at https://%s:%s' % (host, self.port))
+            proto = 'https'
```
Comment (cvium) [inline flexget/webserver.py]: Might as well call it `protocol`.

File: flexget/webserver.py [hunk 4]
```diff
@@ -232,9 +235,11 @@ def _start_server(self):
...
         else:
-            log.info('Web interface available at http://%s:%s' % (host, self.port))
+            proto = 'http'
+
+        log.info('Web interface available at %s://%s:%s' % (proto, host, self.port))
```
Comment (cvium) [inline flexget/webserver.py]: Let the logger handle the string formatting here as well

**Comment by liiight** [comment]
You forgot to change the logger formatting, that's the only thing left

**Comment by b2un0** [comment]
i does not understand what you mean?!

**Comment by liiight** [comment]
Change:
`log.info('Web interface available at %s://%s:%s' % (protocol, host, self.port))`
Into:
`log.info('Web interface available at %s://%s:%s' ,protocol, host, self.port)`
The logger module knows when it needs to interpolate the strings itself and it's considered a good practice.

**Comment by b2un0** [comment]
ah, okay.

i just used the original line and add the protocol parameter;
https://github.com/Flexget/Flexget/blob/develop/flexget/webserver.py#L221

**Comment by liiight** [comment]
yeah, it wasn't right before ;-) might as well do it right the second time.

**Comment by b2un0** [comment]
finally done :)

**Comment by liiight** [comment]
thanks for this

**Pull request merged**